overlord/snapstate/component.go

```diff
@@ -30,8 +31,167 @@ import (
 	"github.com/snapcore/snapd/overlord/state"
 	"github.com/snapcore/snapd/snap"
 	"github.com/snapcore/snapd/snap/naming"
+	"github.com/snapcore/snapd/store"
 )
 
+// InstallComponents installs all of the components in the given names list. The
+// snap represented by info must already be installed, and all of the components
+// in names should not be installed prior to calling this function.
+//
+// TODO:COMPS: respect the transaction that is passed to this function
+func InstallComponents(ctx context.Context, st *state.State, names []string, info *snap.Info, opts Options) ([]*state.TaskSet, error) {
+	var snapst SnapState
+	err := Get(st, info.InstanceName(), &snapst)
+	if err != nil {
+		if errors.Is(err, state.ErrNoState) {
+			return nil, &snap.NotInstalledError{Snap: info.InstanceName()}
+		}
+		return nil, err
+	}
+
+	for _, comp := range names {
+		if snapst.CurrentComponentSideInfo(naming.NewComponentRef(info.SnapName(), comp)) != nil {
+			return nil, snap.AlreadyInstalledComponentError{Component: comp}
+		}
+	}
+
+	compsups, err := componentSetupsForInstall(ctx, st, names, info, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	snapsup := SnapSetup{
+		Base:        info.Base,
+		SideInfo:    &info.SideInfo,
+		Channel:     info.Channel,
+		Flags:       opts.Flags.ForSnapSetup(),
+		Type:        info.Type(),
+		Version:     info.Version,
+		PlugsOnly:   len(info.Slots) == 0,
+		InstanceKey: info.InstanceKey,
+	}
+
+	setupSecurity := st.NewTask("setup-profiles",
+		fmt.Sprintf(i18n.G("Setup snap %q (%s) security profiles"), info.InstanceName(), info.Revision))
+	setupSecurity.Set("snap-setup", snapsup)
+
+	var kmodSetup *state.Task
+	if requiresKmodSetup(&snapst, compsups) {
+		kmodSetup = st.NewTask("prepare-kernel-modules-components", fmt.Sprintf(
+			i18n.G("Prepare kernel-modules components for %q%s"), info.InstanceName(), info.Revision,
+		))
+	}
+
+	tss := make([]*state.TaskSet, 0, len(compsups))
+	compSetupIDs := make([]string, 0, len(compsups))
+	for _, compsup := range compsups {
+		// since we are installing multiple components, we don't want to setup
+		// the security profiles until the end
+		compsup.MultiComponentInstall = true
+
+		componentTS, err := doInstallComponent(st, &snapst, compsup, snapsup, setupSecurity.ID(), setupSecurity, kmodSetup, opts.FromChange)
```

Comment:
are we passing the same tasks along multiple times? what's the real intention there?

Reply:
Yes, we pass the same task in multiple times. See the conversation [here](https://github.com/canonical/snapd/pull/14260#discussion_r1706197490) and [here](https://github.com/canonical/snapd/pull/14260#discussion_r1695648078) for more info.

The intention is to prevent us from having to manually splice the setup-profiles and prepare-kernel-modules-components tasks into the right place in the chain of tasks.

Reply:
it's probably merits a comment as we will have parallel sequences of tasks that converge into the shared one and then split and then merge again etc?